ml-agents/mlagents/trainers/sac/models.py

```diff
@@ -0,0 +1,916 @@
+                q2_heads[name] = _q2
+        q1 = tf.reduce_mean(list(q1_heads.values()), axis=0)
+        q2 = tf.reduce_mean(list(q2_heads.values()), axis=0)
+        # q1 = next(iter(q1_heads.values()))
+        # q2 = next(iter(q2_heads.values()))
+
+        return q1_heads, q2_heads, q1, q2
+
+    def copy_normalization(self, mean, variance, steps):
+        update_mean = tf.assign(self.running_mean, mean)
+        update_variance = tf.assign(self.running_variance, variance)
+        update_norm_step = tf.assign(self.normalization_steps, steps)
+        return tf.group([update_mean, update_variance, update_norm_step])
+
+
+class SACModel(LearningModel):
+    def __init__(
+        self,
+        brain,
+        lr=1e-4,
+        h_size=128,
+        init_entcoef=0.1,
+        max_step=5e6,
+        normalize=False,
+        use_recurrent=False,
+        num_layers=2,
+        m_size=None,
+        seed=0,
+        stream_names=None,
+        tau=0.005,
+        gammas=None,
+        vis_encode_type="default",
+    ):
+        """
+        Takes a Unity environment and model-specific hyper-parameters and returns the
+        appropriate PPO agent model for the environment.
+        :param brain: BrainInfo used to generate specific network graph.
+        :param lr: Learning rate.
+        :param h_size: Size of hidden layers
+        :param init_entcoef: Initial value for entropy coefficient. Set lower to learn faster,
+            set higher to explore more.
+        :return: a sub-class of PPOAgent tailored to the environment.
+        :param max_step: Total number of training steps.
+        :param normalize: Whether to normalize vector observation input.
+        :param use_recurrent: Whether to use an LSTM layer in the network.
+        :param num_layers: Number of hidden layers between encoded input and policy & value layers
+        :param tau: Strength of soft-Q update.
+        :param m_size: Size of brain memory.
+        """
+        self.tau = tau
+        self.gammas = gammas
+        self.brain = brain
+        self.init_entcoef = init_entcoef
+        if stream_names is None:
+            stream_names = []
+        LearningModel.__init__(
+            self, m_size, normalize, use_recurrent, brain, seed, stream_names
+        )
+        if num_layers < 1:
+            num_layers = 1
+        self.last_reward, self.new_reward, self.update_reward = (
+            self.create_reward_encoder()
+        )
+
+        self.policy_network = SACNetwork(
+            brain=brain,
+            m_size=m_size,
+            h_size=h_size,
+            normalize=normalize,
+            use_recurrent=use_recurrent,
+            num_layers=num_layers,
+            seed=seed,
+            stream_names=stream_names,
+            is_target=False,
+            vis_encode_type=vis_encode_type,
+        )
+        self.target_network = SACNetwork(
+            brain=brain,
+            m_size=m_size // 4 if m_size else None,
```

Comment:
Is 4 special here, or should it be an additional parameter?

Reply:
It's special, we want the target network to only have one memory (the policy network has a Policy, Q1, Q2, Value LSTM while the Target only has a Value).